### PR TITLE
[TTS]【Hackathon + No.190】 + 模型复现：iSTFTNet

### LINKED ISSUE #2772: [TTS] iSTFTNet -> speed up HiFiGAN !!
speed up HiFiGAN !!

- Paper: [iSTFTNet : Fast and Lightweight Mel-spectrogram Vocoder Incorporating Inverse Short-time Fourier Transform](https://arxiv.org/pdf/2203.02395.pdf)

- Code: https://github.com/rishikksh20/iSTFTNet-pytorch  

Please compare with original HiFiGAN to get the diff: https://github.com/jik876/hifi-gan 

After this work, you can also modify VITS to VITS_iSTFT

### PR DESCRIPTION
**Pull request by longRookie**
PR types
New features

PR changes
Models

Describe
Same training epochs, iSTFTNet can reduce rtf by more than 50% while ensuring the synthesis quality is not lower than that of HifiGAN.

fix https://github.com/PaddlePaddle/PaddleSpeech/issues/2772

### CODE REVIEW

**Comment by longRookie** [comment]
你好，请问我们的pr存在什么问题吗 @yt605155624

File: examples/csmsc/voc5/iSTFTNet.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
+# iSTFTNet with CSMSC
+
+This example contains code used to train a [iSTFTNet]([[2203.02395] iSTFTNet: Fast and Lightweight Mel-Spectrogram Vocoder Incorporating Inverse Short-Time Fourier Transform (arxiv.org)](https://arxiv.org/abs/2203.02395)) model with [Chinese Standard Mandarin Speech Copus](https://www.data-baker.com/open_source.html),
```
Comment (lym0302) [inline examples/csmsc/voc5/iSTFTNet.md]: 这里展示出来的iSTFTNet 格式不正确，以及最后是句号

File: examples/csmsc/voc5/iSTFTNet.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+```text
+
+iSTFTNet_csmsc_ckpt
+
+├── iSTFT.yaml                    # config used to train iSTFTNet
+
+├── feats_stats.npy               # statistics used to normalize spectrogram when training hifigan
+
+└── snapshot_iter_50000.pdz       # generator parameters of hifigan
```
Comment (lym0302) [inline examples/csmsc/voc5/iSTFTNet.md:128]: 这里显示不对，check 下~

File: examples/csmsc/voc5/conf/iSTFT.yaml [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    post_n_fft: 48                        # stft fft_num
+    gen_istft_hop_size: 12                # istft hop_length
+    gen_istft_n_fft: 48                   # istft n_fft
+    in_channels: 80                       # Number of input channels.
+    out_channels: 1                       # Number of output channels.
+    channels: 512                         # Number of initial channels.
+    kernel_size: 7                        # Kernel size of initial and final conv layers.
+    upsample_scales:  [5,5]               # Upsampling scales.
+    upsample_kernel_sizes: [10,10]        # Kernel size for upsampling layers.
+    resblock_kernel_sizes: [3, 7, 11]     # Kernel size for residual blocks.
```
Comment (lym0302) [inline examples/csmsc/voc5/conf/iSTFT.yaml:37]: 如果想要复用hifigan的话，hifigan 原来的layer 参数 upsample_scales， upsample_kernel_sizes， resblock_kernel_sizes 等均需要保留
所有的地方都需要注意逗号后面有一个空格，代码最后用 pre-commit 刷一下格式

File: examples/csmsc/voc5/conf/iSTFT.yaml [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        - 600000
+        - 800000    
+discriminator_grad_norm: -1             # Discriminator's gradient norm.            
+
+###########################################################
+#                    INTERVAL SETTING                     #
+###########################################################
+generator_train_start_steps: 1     # Number of steps to start to train discriminator.
+discriminator_train_start_steps: 0 # Number of steps to start to train discriminator.
+train_max_steps: 50000          # Number of training steps.
```
Comment (lym0302) [inline examples/csmsc/voc5/conf/iSTFT.yaml]: train_max_steps 应该是要和 hifigan 的配置参数一样

File: examples/csmsc/voc5/conf/iSTFT.yaml [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+window: "hann"           # Window function.
+n_mels: 80               # Number of mel basis.
+fmin: 80                 # Minimum freq in mel basis calculation. (Hz)
+fmax: 7600               # Maximum frequency in mel basis calculation. (Hz)
+
+###########################################################
+#         GENERATOR NETWORK ARCHITECTURE SETTING          #
+###########################################################
+generator_params:
+    istft: True                           # use iSTFTNet
```
Comment (lym0302) [inline examples/csmsc/voc5/conf/iSTFT.yaml]: istft -> use_istft

File: examples/csmsc/voc5/conf/iSTFT.yaml [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+    post_n_fft: 48                        # stft fft_num
+    gen_istft_hop_size: 12                # istft hop_length
+    gen_istft_n_fft: 48                   # istft n_fft
```
Comment (lym0302) [inline examples/csmsc/voc5/conf/iSTFT.yaml]: 这三个参数应该转变为 istft_layer_id，用于表示几层 upsample layer后使用 istft，然后这三个参数应该在代码里面被算出来。这三个参数分别是 istft_n_fft, istft_hop_size, istft_win_size

File: paddlespeech/t2s/models/hifigan/hifigan.py [hunk 7]
```diff
@@ -47,7 +48,12 @@ def __init__(
             nonlinear_activation: str="leakyrelu",
             nonlinear_activation_params: Dict[str, Any]={"negative_slope": 0.1},
             use_weight_norm: bool=True,
-            init_type: str="xavier_uniform", ):
+            init_type: str="xavier_uniform",
+            istft: bool = False,
+            post_n_fft: int=16,
+            gen_istft_hop_size: int=12,
+            gen_istft_n_fft: int=16,
+            ):
```
Comment (lym0302) [inline paddlespeech/t2s/models/hifigan/hifigan.py]: 如 yaml 文件所说修改

File: paddlespeech/t2s/models/hifigan/hifigan.py [hunk 8]
```diff
@@ -79,6 +85,14 @@ def __init__(
...
                 Whether to use weight norm.
                 If set to true, it will be applied to all of the conv layers.
+            istft (bool):
+                If set to true, it will be a iSTFTNet based on hifigan.
+            post_n_fft (int):
+                Emulate nfft in stft
+            gen_istft_hop_size (int):
+                Hop_length in istft
+            gen_istft_n_fft (int):
+                N_fft in istft, equal to post_n_fft
```
Comment (lym0302) [inline paddlespeech/t2s/models/hifigan/hifigan.py]: 同上，对应修改

File: paddlespeech/t2s/models/hifigan/hifigan.py [hunk 9]
```diff
@@ -135,6 +149,13 @@ def __init__(
                 1,
                 padding=(kernel_size - 1) // 2, ),
             nn.Tanh(), )
+        self.istft = istft
+        if self.istft:
+            self.post_n_fft = post_n_fft
+            self.gen_istft_hop_size = gen_istft_hop_size
+            self.gen_istft_n_fft = gen_istft_n_fft
```
Comment (lym0302) [inline paddlespeech/t2s/models/hifigan/hifigan.py]: 这三个参数要注意下命名，参考 yaml 文件的意见，这里需要计算出来数值

File: paddlespeech/t2s/models/hifigan/hifigan.py [hunk 10]
```diff
@@ -135,6 +149,13 @@ def __init__(
...
+            self.reflection_pad = nn.Pad1D(padding=[1,0], mode='reflect')
+            self.conv_post = nn.Conv1D(channels// (2**(i + 1)), self.post_n_fft + 2, 7, 1, padding=3)
```
Comment (lym0302) [inline paddlespeech/t2s/models/hifigan/hifigan.py]: 这里conv的输出维度是计算出来的：self.post_n_fft + 2 -->  (istft_n_fft/2 +1) * 2 , 7 是配置参数中的 kernel_size

File: examples/csmsc/voc5/iSTFTNet.md (hunk 1 above)
Comment (longRookie) [reply examples/csmsc/voc5/iSTFTNet.md]: 好的

File: examples/csmsc/voc5/iSTFTNet.md (hunk 2 above)
Comment (longRookie) [reply examples/csmsc/voc5/iSTFTNet.md:128]: 好的

File: examples/csmsc/voc5/conf/iSTFT.yaml (hunk 3 above)
Comment (longRookie) [reply examples/csmsc/voc5/conf/iSTFT.yaml:37]: 已经修改为和hifigan一致，通过istft_layer_id参数决定upsample layer层数；代码已经使用pre-commit格式化

File: examples/csmsc/voc5/conf/iSTFT.yaml (hunk 4 above)
Comment (longRookie) [reply examples/csmsc/voc5/conf/iSTFT.yaml]: 已经修改为2500000，和hifigan配置参数一样

File: examples/csmsc/voc5/conf/iSTFT.yaml (hunk 5 above)
Comment (longRookie) [reply examples/csmsc/voc5/conf/iSTFT.yaml]: 已修改

File: examples/csmsc/voc5/conf/iSTFT.yaml (hunk 6 above)
Comment (longRookie) [reply examples/csmsc/voc5/conf/iSTFT.yaml]: 这里yaml更改为三个参数：
* `use_istft`：表示是否使用iSTFTNet，如果为False等同于hifigan;
* `istft_layer_id`： 在第 `istft_layer_id` 个 上采样层后使用istft;
* `overlap_ratio`：表示重叠率（overlap ratio）或者叫做帧移率（hop size ratio），等于n_fft/hop_size，默认为4（相比于设置为1和2，在设置为4时，实验的合成效果更好），参考[paddle istft  hop_length](https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/signal/istft_cn.html#canshu)；

 `istft_hop_size`在代码里面计算，`istft_n_fft`为`overlap_ratio`*`istft_hop_size`，`istft_win_size`设置为等于`istft_n_fft`

File: paddlespeech/t2s/models/hifigan/hifigan.py (hunk 7 above)
Comment (longRookie) [reply paddlespeech/t2s/models/hifigan/hifigan.py]: 这里对应yaml的修改：
` use_istft: bool=False,
            istft_layer_id: int=2,
            overlap_ratio: float=4,`
https://github.com/longRookie/PaddleSpeech/blob/e91bff79f56b57a80cf10755ac10192bb5a3e5ff/paddlespeech/t2s/models/hifigan/hifigan.py#L51

File: paddlespeech/t2s/models/hifigan/hifigan.py (hunk 8 above)
Comment (longRookie) [reply paddlespeech/t2s/models/hifigan/hifigan.py]: 已对应修改

File: paddlespeech/t2s/models/hifigan/hifigan.py (hunk 9 above)
Comment (longRookie) [reply paddlespeech/t2s/models/hifigan/hifigan.py]: 已修改
```python
        if self.use_istft:
            self.istft_hop_size = 1
            for j in range(istft_layer_id, len(upsample_scales)):
                self.istft_hop_size *= upsample_scales[j]
            self.istft_layer_id = istft_layer_id
            self.istft_n_fft = int(self.istft_hop_size * overlap_ratio)
            self.istft_win_size = self.istft_n_fft
```

File: paddlespeech/t2s/models/hifigan/hifigan.py (hunk 10 above)
Comment (longRookie) [reply paddlespeech/t2s/models/hifigan/hifigan.py]: 已修改
```python
            self.conv_post = nn.Conv1D(
                channels // (2**(i + 1)),
                (self.istft_n_fft // 2 + 1) * 2,
                kernel_size,
                1,
                padding=(kernel_size - 1) // 2, )
```

Comment (longRookie) [reply paddlespeech/t2s/models/hifigan/hifigan.py]: 我点错了，/(ㄒoㄒ)/~~

File: paddlespeech/t2s/models/hifigan/hifigan.py (hunk 9 above)
Comment (longRookie) [reply paddlespeech/t2s/models/hifigan/hifigan.py]: 我点错了，对不起😭

File: examples/csmsc/voc5/conf/iSTFT.yaml (hunk 6 above)
Comment (lym0302) [reply examples/csmsc/voc5/conf/iSTFT.yaml]: overlap_ratio：表示重叠率（overlap ratio）或者叫做帧移率（hop size ratio），等于n_fft/hop_size，默认为4（相比于设置为1和2，在设置为4时，实验的合成效果更好），参考[paddle istft hop_length](https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/signal/istft_cn.html#canshu)  这个是经过实验的吗？整体用这一套参数是训练出模型OK的吗？因为和论文里的算法不太一样，所以需要和你确认下~

File: paddlespeech/t2s/models/hifigan/hifigan.py [hunk 11]
```diff
@@ -135,6 +147,23 @@ def __init__(
...
+        self.use_istft = use_istft
+        if self.use_istft:
+            self.istft_hop_size = 1
+            for j in range(istft_layer_id, len(upsample_scales)):
+                self.istft_hop_size *= upsample_scales[j]
+            self.istft_layer_id = istft_layer_id
+            self.istft_n_fft = int(self.istft_hop_size * overlap_ratio)
+            self.istft_win_size = self.istft_n_fft
+            self.reflection_pad = nn.Pad1D(padding=[1, 0], mode='reflect')
+            self.conv_post = nn.Conv1D(
```
Comment (lym0302) [inline paddlespeech/t2s/models/hifigan/hifigan.py]: self.output_conv 可以写到 if self.use_istft 和 else 中。即 把 141~ 149行移到166 行，然后这里 self.conv_post 该名成 self.output_conv

File: paddlespeech/t2s/models/hifigan/hifigan.py [hunk 12]
```diff
@@ -135,6 +147,23 @@ def __init__(
...
+                channels // (2**(i + 1)),
+                (self.istft_n_fft // 2 + 1) * 2,
+                kernel_size,
+                1,
+                padding=(kernel_size - 1) // 2, )
+        else:
+            self.istft_layer_id = len(upsample_scales)
```
Comment (lym0302) [inline paddlespeech/t2s/models/hifigan/hifigan.py]: 如果 forword 的时候没有用到 self.istft_layer_id 的话，可以把这个去掉

File: examples/csmsc/voc5/conf/iSTFT.yaml (hunk 6 above)
Comment (longRookie) [reply examples/csmsc/voc5/conf/iSTFT.yaml]: image

我们复现的模型为C8C8I；
在https://github.com/rishikksh20/iSTFTNet-pytorch/blob/ecbf0f635b36432bd3e432790326591bc86cadbc/config_v1.json#L21 中同样采用的是hop_size=4,  "n_fft": 1024,"hop_size": 256；

根据[这三个参数应该在代码里面被算出来](https://github.com/PaddlePaddle/PaddleSpeech/pull/3006#discussion_r1144327337)，我们将hop_size和n_fft设置为在代码中计算，overlap_ratio取4，来和论文中设置一致；

另外，我们也尝试了让ratio= 1, 2，合成的音质会有损失，我们认为这是因为nfft的减小带来了频率分辨率的降低影响了合成的音质。
在ratio=4时，具体得到的hop_size和nfft和论文不一致，这是因为paddlespeech中hifigan的采用upsample_rates和原始hifigan中不同，paddlespeech中为[5,5,4,3](https://github.com/PaddlePaddle/PaddleSpeech/blob/d3d86f59aa3d9bc56bdc5f355350ba019e6e7e99/examples/csmsc/voc5/conf/default.yaml#L31)，hifigan中为[8,8,2,2](https://github.com/jik876/hifi-gan/blob/4769534d45265d52a904b850da5a622601885777/config_v1.json#L11)，如果后两个元素的乘积为4，我们的hop_size和n_fft将和原论文一致，

训练模型是ok的，我们在对应的[iSTFTNet.md](https://github.com/longRookie/PaddleSpeech/blob/e91bff79f56b57a80cf10755ac10192bb5a3e5ff/examples/csmsc/voc5/iSTFTNet.md)中给出了50000次的hifigan和istftNet的预训练模型百度网盘链接和实验结果对比

File: paddlespeech/t2s/models/hifigan/hifigan.py (hunk 12 above)
Comment (longRookie) [reply paddlespeech/t2s/models/hifigan/hifigan.py]: 已经修改

Comment (longRookie) [reply paddlespeech/t2s/models/hifigan/hifigan.py]: 已经修改

File: paddlespeech/t2s/models/hifigan/hifigan.py (hunk 11 above)
Comment (longRookie) [reply paddlespeech/t2s/models/hifigan/hifigan.py]: 已经修改，为了使得之前训练好的iSTFTNet模型可以在将self.conv_post修改为self.output_conv之后正常调用。
更新了模型的state_dict使得和修改后的模型匹配，已测试通过正常合成；
具体方法：
```python
    state_dict = paddle.load(args.checkpoint)
    generator_params = state_dict['generator_params']
    print(generator_params.keys())
    keys_list = list(generator_params.keys())
    for key in keys_list:
        if "conv_post" in key:
            old_key = key
            new_key = key.replace("conv_post","output_conv")
            generator_params[new_key] = generator_params.pop(old_key)
        if "output_conv" in key:
            del generator_params[key]

    paddle.save(state_dict,'/home/aistudio/PaddleSpeech/examples/csmsc/voc5/istft_snapshot_iter_50000.pdz')
```

File: examples/csmsc/voc5/conf/iSTFT.yaml (hunk 6 above)
Comment (lym0302) [reply examples/csmsc/voc5/conf/iSTFT.yaml]: 论文中的计算方式是如下图计算，我们的 f1, h1,w1和论文的不一样，论文中的设置是刚好满足 ratio=4，我们的根据下述公式的话是不存在 ratio=4的，而且 ratio=4 我感觉不是必须的，只要 win_size > hop_size 就行， hop_size 和 fft 的关系也不一定是 ratio 的关系。我的疑惑点是这个
image

Comment (lym0302) [reply examples/csmsc/voc5/conf/iSTFT.yaml]: 我们的 hifigan 提取特征的配置是，按照你的算法 ，overlap_ratio =  2048 / 300
image

**Review by lym0302 — APPROVED** [review]
